### ISSUE TITLE
Why is ExecuteScript API so slow ? Are there any ways to make it faster?

### ISSUE THREAD

**Issue by alan008** [MicrosoftEdge/WebView2Feedback#2682]
I try to use WebView2 from Delphi.

In Delphi there is a TLB wrapper, in which ExecuteScript is described as follows:
    function ExecuteScript(javaScript: PWideChar; 
                          const handler: ICoreWebView2ExecuteScriptCompletedHandler): HResult; stdcall;
When I call ExecuteScript 1000 times in cycle , even with the simplest script, returning just string "OK" and doing nothing else, I can't call ExecuteScript more than 250 times per second (I always wait for CompletedHandler to execute before I call ExecuteScript again, because I need to execute scripts in synchronous manner, result from previous script execution is needed for next script). I think 250 times per second is too slow? Or not? It there any way to call ExecuteScript faster?

**Version**
SDK: 
Runtime: Microsoft.WebView2.FixedVersionRuntime.103.0.1264.71.x64.cab, Microsoft.WebView2.FixedVersionRuntime.103.0.1264.71.x86.cab
Framework: VCL(WinForms) in Delphi 
OS: Win10

**Comment by champnic** [comment]
Hey @alan008 - I'm not sure. I've added our perf expert who may be able to weigh in here. Is there any way you can batch the calls or otherwise require fewer calls to ExecuteScript or between the app and the web? It is a cross-process call from the app to the browser process.

**Comment by LiangTheDev** [comment]
That sounds about right to me. It involves getting the request over to renderer process, compile and run the script and then return the result. It is likely that we could not get it to run faster.
You can try postWebMessage to the page, and let a handler on the page to react to the message and do the calling of script and then post the result as a web message back. That might be faster. The reason is that we don't have to compile the script every time. If there are some functions got called repeatedly, the native code should have already been generated after several calls.

**Comment by alan008** [comment]
The most strange thing is that, despite very slow execution, CPU usage is very low (even on a single core) - at about 5%. So the strange thing is where it hangs? Could you please do some profiling on your side?

**Comment by alan008** [comment]
Oh, and one more strange. The speed not depends on the JS code I'm executing. I can send different code to execute script - more simple or more complex code, but execution speed is always at about 250 calls per second. So the slow part is not at the execution itself. Maybe it is on some preparation/compiling or something, but even if I do ExecuteScript("OK"), it is also 250 times per second, and no "speedup" effect from pre-compilation or something even when this is called in cycle 250 times one call after another.

**Comment by LiangTheDev** [comment]
If it is always 250 times for simple or complex script, then it sounds like something else is in play.

WebView2 runs in separate process and there are interprocess communications involved. The ExecuteScript result goes through message pump of the thread where WebView is running. Not saying that it is the case, if the message pump only pumps 250 messages per second, then it could be a limiter.

Things to try out:
- Since you mentioned "even on a single core", I assume that you are testing in a VM. Have you tested on a physical machine?
- Have you tested in sample WPF applications at https://github.com/MicrosoftEdge/WebView2Samples? I don't think that we have this strange 250 calls per sound regardless of script complexity issue with sample app. If it still repros, then it might be some machine level setup.
- If we still could not find any clue, we might have to take a performance trace following instruction at https://github.com/MicrosoftEdge/WebView2Feedback/issues/1540#issuecomment-892407347. We don't have Delphi environment, your machine setup or specific app code. So, it would be better for you to take a trace and then share with us. The trace file could be large, so we would have to share it through some cloud drive.

**Comment by alan008** [comment]
Great thanks for help and involment into my problem!

No, I don't use VM, I test on physical machine with modern Core i5. My message about 1 core utilization was to emphasize, that there was no CPU utilization on ANY of cores, most cores were idle, when one or two cores were at about 5% of load.

Here is my minimal Delphi example project, sorry for 160+ MB download: I've included fixed version runtime so you can test it "just like on my machine".
https://www.dropbox.com/s/z9l8vxyqvzkq1hj/ExecuteScriptPerf.zip?dl=1

In folder bin\Debug theres 32 bit app Project1.exe, you can run ant test. Runtime is located also in bin. How much calls per second do you get?

**Comment by alan008** [comment]
Arrgh! My mistake - I used Sleep(1) insted of Sleep(0) waiting for result and doing ProcessMessages in cycle, but I forgot, that sleep(1) actually can't sleep 1 millisecod, it sleeps at around 15-20 milliseconds due to system timer low precision and thread quantization. Changed to Sleep(0) for now and got more than x10 performance. 
Will think about something like 
if Random(100)=1 then 
  Sleep(0) 
to sleep even less.
or I'll use counter and will sleep only 1 time per 100 cycles

I think ticket could be closed now. Thanks for all the help!
